File: 2020/day_10/day_10.py
from collections import defaultdict

with open('in') as f:
    data = list(map(lambda x: int(x.strip()), f.readlines()))
    data.append(0)
    data.append(max(data) + 3)
    data.sort()


def part_one():
    diffs = defaultdict(int)

    for i in range(len(data) - 1):
        diffs[data[i + 1] - data[i]] += 1

    return diffs[1] * diffs[3]


def part_two():
    memo = dict()

    def solver(start_index):
        if start_index == len(data) - 1:
            return 1

        if start_index in memo:
            return memo[start_index]

        count = 0
        for i in range(start_index + 1, len(data)):
            if data[i] - data[start_index] <= 3:
                count += solver(i)
            else:
                break

        memo[start_index] = count
        return count

    return solver(0)


if __name__ == '__main__':
    print(f'Part One: {part_one()}')
    print(f'Part Two: {part_two()}')

File: 2020/day_10/day_10.rs
use std::fs;
use std::collections::HashMap;

fn main() {
    let f = fs::read_to_string("in").unwrap();
    let mut data: Vec<i32> = f.lines().map(|x| x.trim().parse().unwrap()).collect();
    data.push(0);
    data.push(*data.iter().max().unwrap() + 3);
    data.sort();

    println!("Part One: {}", part_one(&data));
    println!("Part Two: {}", part_two(&data));
}

fn part_one(data: &Vec<i32>) -> i32 {
    let (diff1, diff3) = data
        .iter()
        .skip(1)
        .zip(data.iter())
        .map(|(x, y)| x - y)
        .fold((0, 0), |(diff1, diff3), x| {
            if x == 1 {
                (diff1 + 1, diff3)
            } else {
                (diff1, diff3 + 1)
            }
        });

    diff1 * diff3
}

fn part_two(data: &Vec<i32>) -> u64 {
    let mut memo = HashMap::new();
    part_two_solver(data, 0, &mut memo)
}

fn part_two_solver(data: &Vec<i32>, index: usize, memo: &mut HashMap<usize, u64>) -> u64 {
    if index == data.len() - 1 {
        return 1;
    }

    if let Some(res) = memo.get(&index) {
        return *res;
    }

    let mut count = 0;
    for i in (index+1)..data.len() {
        if data[i] - data[index] <= 3{
            count += part_two_solver(data, i, memo);
        } else {
            break;
        }
    }

    memo.insert(index, count);
    count
}
